leakpro/signals/utils/HopSkipJumpDIstance.py

```diff
@@ -0,0 +1,794 @@
+        elif self.constraint == np.inf:
+            projected = clamp(
+                perturbed_images_batch, original_image_batch - alphas, original_image_batch + alphas
+            )
+        return projected
+
+
+    def geometric_progression_for_stepsize(self: Self,
+                                           samples: Tensor,
+                                           updates: Tensor,
+                                           dist: Tensor,
+                                           current_iteration: int,
+                                           active_indices: np.ndarray, b_i:int) -> Tensor:
+        """Calculates the step size for geometric progression in the HopSkipJumpDistance algorithm.
+
+        Args:
+        ----
+            samples (Tensor): The input samples.
+            updates (Tensor): The updates to be applied to the samples.
+            dist (Tensor): The distance value.
+            current_iteration (int): The current iteration of the algorithm.
+            active_indices (np.ndarray): The active indices.
+            b_i (int): The value of b_i.
+
+        Returns:
+        -------
+            Tensor: The calculated step size for geometric progression.
+
+        """
+        batch_epsilon = dist / np.sqrt(current_iteration)
+        active_indices = np.array(active_indices)
+
+        success = np.zeros(len(samples), dtype=bool)
+        batch_active_indices = np.arange(len(samples))
+        num_evals = 0
+
+        while not success.all() and num_evals <= 10:
+
+            active_samples = samples[batch_active_indices]
+            active_updates = updates[batch_active_indices]
+            active_epsilon = batch_epsilon[batch_active_indices].view(len(batch_active_indices), 1, 1, 1)
+
+            active_updateed_samples = active_samples + active_epsilon * active_updates
+            decisions, _  = self.decision_function(active_updateed_samples,
+                                                   active_indices[batch_active_indices])
+            passed_index= []
+            for i, idx in enumerate(batch_active_indices):
+                if decisions[i]:
+                    success[idx] = True
+                    passed_index.append(idx)
+                else:
+                    batch_epsilon[idx] =  batch_epsilon[idx] / 2.0
+            if passed_index:
+                # Find the positions in batch_active_indices corresponding to the values in passed_index
+                positions_to_delete = [i for i, val in enumerate(batch_active_indices) if val in passed_index]
+
+                batch_active_indices = np.delete(batch_active_indices, positions_to_delete)
+            num_evals += 1
+            if self.verbose:
+                message = f"Stepsize search: {num_evals} evaluations, and failed data {len(batch_active_indices)} in {b_i}."
+                self.logger.info(message)
+        return batch_epsilon
```

Comment:
the return does not entail what samples were successful. Isnt this information that is useful to us? Should we really attack a point we didnt succeed with?